### DISCUSSION TITLE
Dashboard question

### DISCUSSION THREAD

**Discussion by tomy2105** [elsa-workflows/elsa-core#989] [Q&A] [answered]
Items in dashboard:
Workflow Definitions - where you edit workflows in designer
Workflow Instances - "history" of workflow executions

Could you please explain what does Workflow Registry in dashboard show?

And should it have create button? Because it does same thing as Workflow Definitions create button? A bit confusing :(.

**Comment by sfmskywalker** [comment, marked as answer]
**Workflow Definition** 

A "dynamic" workflow model that is created by the workflow designer (or hand-craft using JSON, which is how the workflow definition is serialized ultimately).

When the application starts, Elsa collects all available *Workflow Blueprints* from the *Workflow Registry*. This workflow registry requests workflow blueprints from *workflow providers*. 

One such workflow provider is the `DatabaseWorkflowProvider` which reads all *workflow definitions* and maps them into *workflow blueprints*.

> Similarly, there is a `ProgrammaticWorkflowProvider` which resolves all `IWorkflow` implementations (representing coded workflow classes using the Workflow Builder API) and turns them into workflow blueprints too.

**Workflow Registry**

The Workflow Registry is a collection of *Workflow Blueprints*, which are retrieved through a collection of `IWorkflowProvider` implementations, currently comprised of:

- `DatabaseWorkflowProvider`
- `DatabaseWorkflowProvider`
- `StorageWorkflowProvider` (similar to DatabaseWorkflowProvider, but loads JSON files instead)

As you can see, Elsa uses various sources to build up a full registry of workflow blueprints, which are displayed on the Elsa Dashboard.
Workflow Definitions are just one source from which Workflow Blueprints are constructed. Programmatic workflows (workflow classes) are another source.

Workflow Blueprints themselves aren't editable through the Dashboard, which is why you can edit Workflow Definitions but not when viewed from the Workflow Registry screen.

That being said, I do have something in mind that unifies the concept of Workflow Definition a bit, which also solves the issue of programmatic workflow versioning (as described in #569).

With that change, Workflow Definitions will be promoted to being able to carry any type of content: JSON as well as binary assemblies.

With or without this change, we can dynamically display an "edit" button on workflow blueprints on the workflow registry screen if the workflow blueprint originated from an "editable" source (such as workflows created from the designer).

Further down the rabbit hole, we might even consider another update where we allow the user to code workflows using C# straight from the designer, since #569 would enable us to compile C# files into assemblies ourselves easily. Then even for programmatic workflows we could show an "edit" button.

**Reply by tomy2105** [reply]
Understood the difference now. Thanks.

**Reply by asomwanshi37** [reply]
Hello @sfmskywalker  can I Integrate Elsa Dashboard??????
